cache/restore.go

```diff
@@ -0,0 +1,164 @@
+	downloadTime := time.Since(downloadStartTime).Round(time.Second)
+	r.logger.Donef("Downloaded archive in %s", downloadTime)
+	tracker.logArchiveDownloaded(downloadTime, fileInfo, len(config.Keys))
+
+	r.logger.Println()
+	r.logger.Infof("Restoring archive...")
+	extractionStartTime := time.Now()
+	if err := compression.Decompress(archivePath, r.logger, r.envRepo); err != nil {
+		return fmt.Errorf("failed to decompress cache archive: %w", err)
+	}
+	extractionTime := time.Since(extractionStartTime).Round(time.Second)
+	r.logger.Donef("Restored archive in %s", extractionTime)
+	tracker.logArchiveExtracted(extractionTime, len(config.Keys))
+	tracker.wait()
+
+	return nil
+}
+
+func (r *restorer) createConfig(input RestoreCacheInput) (restoreCacheConfig, error) {
+	apiBaseURL := r.envRepo.Get("BITRISEIO_ABCS_API_URL")
+	if apiBaseURL == "" {
+		return restoreCacheConfig{}, fmt.Errorf("the secret 'BITRISEIO_ABCS_API_URL' is not defined")
+	}
+	apiAccessToken := r.envRepo.Get("BITRISEIO_ABCS_ACCESS_TOKEN")
+	if apiAccessToken == "" {
+		return restoreCacheConfig{}, fmt.Errorf("the secret 'BITRISEIO_ABCS_ACCESS_TOKEN' is not defined")
+	}
+
+	keys, err := r.evaluateKeys(input.Keys)
+	if err != nil {
+		return restoreCacheConfig{}, fmt.Errorf("failed to evaluate keys: %w", err)
+	}
+
+	return restoreCacheConfig{
+		Verbose:        input.Verbose,
+		Keys:           keys,
+		APIBaseURL:     stepconf.Secret(apiBaseURL),
+		APIAccessToken: stepconf.Secret(apiAccessToken),
+	}, nil
+}
+
+func (r *restorer) evaluateKeys(keys []string) ([]string, error) {
+	model := keytemplate.NewModel(r.envRepo, r.logger)
+	buildContext := keytemplate.BuildContext{
+		Workflow:   r.envRepo.Get("BITRISE_TRIGGERED_WORKFLOW_ID"),
+		Branch:     r.envRepo.Get("BITRISE_GIT_BRANCH"),
+		CommitHash: r.envRepo.Get("BITRISE_GIT_COMMIT"),
+	}
+
+	var evaluatedKeys []string
+	for _, key := range keys {
+		if key == "" {
+			continue
+		}
+
+		r.logger.Println()
+		r.logger.Printf("Evaluating key template: %s", key)
+		evaluatedKey, err := model.Evaluate(key, buildContext)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate key template: %s", err)
+		}
+		r.logger.Donef("Cache key: %s", evaluatedKey)
+		evaluatedKeys = append(evaluatedKeys, evaluatedKey)
+	}
+
+	return evaluatedKeys, nil
+}
+
+func (r *restorer) download(config restoreCacheConfig) (string, error) {
+	dir, err := os.MkdirTemp("", "restore-cache")
+	if err != nil {
+		return "", err
+	}
+	name := fmt.Sprintf("cache-%s.tzst", time.Now().UTC().Format("20060102-150405"))
+	downloadPath := filepath.Join(dir, name)
+
+	params := network.DownloadParams{
+		APIBaseURL:   string(config.APIBaseURL),
+		Token:        string(config.APIAccessToken),
+		CacheKeys:    config.Keys,
+		DownloadPath: downloadPath,
+	}
+	err = network.Download(params, r.logger)
+	if err != nil {
+		if errors.Is(err, network.ErrCacheNotFound) {
+			r.logger.Donef("No cache entry found for the provided key")
+			os.Exit(0)
```

Comment:
Is this right?

Reply:
Yes, this is the intended behavior. Caching is optional, and marking the step as "failed" just because there is no stored cache is confusing and we received feedback about it for the old cache steps, which triggered this other PR: https://github.com/bitrise-steplib/steps-cache-pull/pull/38

Reply:
As I understood it could be used from other Steps, are you sure we do not want to do other work after this?
For example we may want to still run the export part of the step architecture.
In this case we could just return nil.